zerver/worker/queue_processors.py

```diff
@@ -600,24 +597,30 @@ def maybe_send_batched_emails(self) -> None:
             self.timer_event = None
 
             current_time = time.time()
-            for user_profile_id, timestamp in list(self.batch_start_by_recipient.items()):
-                if current_time - timestamp < self.BATCH_DURATION:
+
+            all_user_batches = MissedMessageEmailBatch.objects.all().select_related()
+            for user_batch in all_user_batches:
+                user_profile_id: int = user_batch.user_profile_id
+                events: List[Dict[str, Any]] = user_batch.events
+
+                next_process_time: datetime.datetime = user_batch.next_process_time
+                if current_time - time.mktime(next_process_time.timetuple()) < self.BATCH_DURATION:
                     continue
-                events = self.events_by_recipient[user_profile_id]
+
                 logging.info(
                     "Batch-processing %s missedmessage_emails events for user %s",
                     len(events),
                     user_profile_id,
                 )
                 handle_missedmessage_emails(user_profile_id, events)
-                del self.events_by_recipient[user_profile_id]
-                del self.batch_start_by_recipient[user_profile_id]
+                user_batch.delete()
 
             # By only restarting the timer if there are actually events in
             # the queue, we ensure this queue processor is idle when there
             # are no missed-message emails to process.  This avoids
             # constant CPU usage when there is no work to do.
-            if len(self.batch_start_by_recipient) > 0:
+            # Faster than len()
+            if MissedMessageEmailBatch.objects.count() > 0:
```

Comment:
Should this be `.exists()`?

Reply:
I suppose we could actually avoid this query entirely by just keeping track of whether we did a `continue` above.

Reply:
Changed to `exists` 👍